### ISSUE TITLE
OpCopyLogical should do bool<->uint8_t conversions

### ISSUE THREAD

**Issue by seanbaxter** [KhronosGroup/SPIRV-Headers#181]
The purpose of OpCopyLogical is to convert between layout OpTypeStructs (eg for Uniform and StorageBuffer storage classes) and non-layout OpTypeStructs (eg for Function and Private storage classes). Since OpTypeBool isn't a layout type, layouts need to store these as uint8_t and OpINotEqual to get a bool back out. This should be a supported operation of OpCopyLogical, because without it, it won't actually perform layout<->non-layout conversions in a general way.

**Comment by johnkslang** [comment]
But, it is not formalized in SPIR-V what the physical representation of a Boolean is.  E.g., Vulkan uses 32-bit int for storage buffers.  I think something needs to remain abstracted here.  Or, possibly parameterized, to avoid resorting to member-wise copy when boolean conversions are involved.

**Comment by seanbaxter** [comment]
OpCopyLogical should do bool<->any integral type. Both types involved in the conversion are explicit, so there's no chance of the operator favoring one boolean representation over another.

**Comment by csyonghe** [comment]
Agreed. OpCopyLogical isn’t going to be that useful without allowing bool to int conversion.

This issue is particularly acute if the user define something like bool[8192] in uniform buffer. Without it we have no choice but to generate a for loop to do the conversion to load the array into a Function local variable to avoid generating a ton of bloated code. But that loop is typically not going to be optimized well by drivers.

More generally for the load direction, I think Vulkan should remove its validation rule to force Function variables to have types that are not explicitly laid out and instead just allow local variables whose type is a std140 laid out struct/array to avoid this kind of unnecessary copies.
